src/Migration/Service/MediaFileProcessorService.php

```
@@ -45,8 +45,10 @@ public function processMediaFiles(MigrationContextInterface $migrationContext, C
             if ($currentDataSet === null) {
                 try {
                     $currentDataSet = $this->dataSetRegistry->getDataSet($migrationContext, $mediaFile['entity']);
-                } catch (DataSetNotFoundException $exception) {
-                    $this->logDataSetNotFoundException($migrationContext, $exception);
+                } catch (MigrationException $exception) {
+                    if ($exception->getErrorCode() === MigrationException::DATASET_NOT_FOUND) {
```

Same applies to this change, but slightly differently:

Previously every other Exception that wasn't `DataSetNotFoundException` was still thrown / bubble up. Now all kinds of `MigrationException` that don't have that specific error code are silently ignored, which is very bad. Sure you could add an else branch and throw it again, but it's just isn't as clean as a proper dedicated exception.

In general every time we explicitly need to catch an Exception, it totally makes sense to have a dedicated Exception class (which could still inherit from the Domain Exception). Btw, that is also documented here 🙂 : https://developer.shopware.com/docs/resources/references/adr/2022-02-24-domain-exceptions.html

---

The function called here can really only throw this exception and depth is also just one layer, so we could also just get rid of the if check and log if a exception is thrown... don't think that with this less complexity a own Exception class is necessary

---

Even if that is the case today, I still think it's a code smell, because it's very easy to introduce a bug (with your updated version emitting a log that is misleading).

For example if someone is working on the class in the try block, for whatever reason throwing another type of `MigrationException`, he can't safely assume that it will be properly handled in all places and would need to adjust this try block as well / has to know the whole codebase.

If it would be a dedicated exception, it's much clearer.

I'm totally aware that it is very unlikely that someone would change these Registries and suddenly throw a different exception. But on the other hand having multiple try / catch blocks after one another also isn't very clean. I would still vote for leaving a dedicated exception for these cases.

---

but with this argument we can't use Domain Exceptions, actually any Exception or try-catch anywhere. Yeah when somebody changes the Exception in the class "below" where its thrown and not the catch then we have the same situation, also with any other Exception.

---

That's not totally true, in other places you might catch our generic domain exception and deal with it like "something gone wrong with the migration", e.g. by creating a proper log entry.

But in this specific case you are only interested in one very specific exception (that the registry can't find something) and are handling it exactly like that, but still catching a generic `MigrationException`, which could be many things.

If you would deal with the `MigrationException` here just like "something gone wrong with the migration" and emitting a generic error log, everything would be totally fine. I just don't like that it's catching a `MigrationException` and then always automatically assumes it must be that the registry couldn't find the item, which might not always be true.

And that is exactly why we still use dedicated exception classes next to our domain exceptions, when they make sense / are catched in specific places.

---

I got that part, but what speaks agains `if ($exception->getErrorCode() === MigrationException::PROCESSOR_NOT_FOUND)` then?

---

The missing `else` block 😅. In this specific place an `else { throw $e; }` would have been also fine, it just makes sure the exception isn't "silenced"

Additionally it's not as clean to read as multiple / a more dedicated catch block. But that's only a nitpick